src/psm/src/ir_solver.cpp

```diff
@@ -654,81 +653,120 @@ void IRSolver::createGmatNodes(const vector<dbSBox*>& power_wires,
+      y_loc1 = curWire->yMin();
+      y_loc2 = curWire->yMax();
+    }
+    // For all layers we create the end nodes
+    Gmat_->setNode({x_loc1, y_loc1}, l);
+    Gmat_->setNode({x_loc2, y_loc2}, l);
+    // Special condition: if the stripe ovelaps a macro ensure a node is
+    // created
+    for (const auto& macro : macros) {
       if (layer_dir == dbTechLayerDir::Value::HORIZONTAL) {
-        y_loc1 = (curWire->yMin() + curWire->yMax()) / 2;
-        y_loc2 = y_loc1;
-        x_loc1 = curWire->xMin();
-        x_loc2 = curWire->xMax();
+        // y range is withing the marco (min, max)
+        if (y_loc1 >= macro.yMin() && y_loc1 <= macro.yMax()) {
+          // Both x values outside the macro
+          // (Values inside will already have a node at endpoints)
+          if (x_loc1 < macro.xMin() && x_loc2 > macro.xMax()) {
+            const int x = (macro.xMin() + macro.xMax()) / 2;
+            Gmat_->setNode({x, y_loc1}, l);
+          }
+        }
       } else {
-        x_loc1 = (curWire->xMin() + curWire->xMax()) / 2;
-        x_loc2 = x_loc1;
-        y_loc1 = curWire->yMin();
-        y_loc2 = curWire->yMax();
+        if (x_loc1 >= macro.xMin() && x_loc1 <= macro.xMax()) {
+          if (y_loc1 < macro.yMin() && y_loc2 > macro.yMax()) {
+            const int y = (macro.yMin() + macro.yMax()) / 2;
+            Gmat_->setNode({x_loc1, y}, l);
+          }
+        }
       }
-      // special case for bottom layers we design a dense grid at a fixed
-      // frequency
-      if (l == bottom_layer_) {
+    }
+    if (l != bottom_layer_)
+      continue;
+
+    // special case for bottom layers we design a dense grid at a fixed
+    // frequency
+    auto node_map = Gmat_->getNodes(l,layer_dir,x_loc1,x_loc2,y_loc1,y_loc2);
+    pair<pair<int, int>, Node*> node_prev;
+    int v_itr, v_prev, length;
+    int i = 0;
+    for (auto& node_itr : node_map) {
+      v_itr = (node_itr.first).first;
+      if (i == 0) {
+        // Before the first existing node
+        i = 1;
         if (layer_dir == dbTechLayerDir::Value::HORIZONTAL) {
-          // quantize the horizontal direction
-          x_loc1 = (x_loc1 / node_density_) * node_density_;
-          x_loc2 = (x_loc2 / node_density_) * node_density_;
-          for (int x_i = x_loc1; x_i <= x_loc2; x_i += node_density_) {
-            Gmat_->setNode({x_i, y_loc1}, l);
-          }
+          v_prev= x_loc1;
         } else {
-          // quantize the vertical direction
-          y_loc1 = (y_loc1 / node_density_) * node_density_;
-          y_loc2 = (y_loc2 / node_density_) * node_density_;
-          for (int y_i = y_loc1; y_i <= y_loc2; y_i += node_density_) {
-            Gmat_->setNode({x_loc1, y_i}, l);
+          v_prev= y_loc1;
+        }
+      } else {
+        v_prev = (node_prev.first).first;
+      }
+      length = v_itr - v_prev;
+      if (length > node_density_) {
+        for (int v_i = v_prev+ node_density_; v_i < v_itr;
+               v_i += node_density_) {
```

Comment:
It would be better to even distribute the nodes over the gap.  For example if density is 5 and the nodes are at 0 and 6 it would be better to add a node at 3 instead of 5.

Reply:
Thanks resolved.